### PR TITLE
Create new S3 buckets for OCW Pipeline End to End Test

### LINKED ISSUE #1945: create S3 buckets for OCW end to end testing pipeline
# Description/Context
With https://github.com/mitodl/ocw-studio/pull/2018, we will add a new end to end testing pipeline definition to `ocw-studio`. This pipeline will build 2 test courses, one `ocw-www` site (aka the root website) and one course site. These sites will then be deployed to S3 and have Playwright run against them. This allows us to run our end to end tests against sites built using the exact same steps as sites published through the UI. 

# Plan/Design
In order for this to work, we will need a total of 4 new S3 buckets. Two of these (the online versions) will need Fastly distributions pointed at them. Ideally, it would look something like this:

 - `ocw-content-test-qa` -> `https://test-qa.ocw.mit.edu/`
 - `ocw-content-offline-test-qa` (no Fastly)
 - `ocw-content-test-production` -> `https://test.ocw.mit.edu/`
 - `ocw-content-offline-test-production` (no Fastly)

The offline buckets would be created pre-emptively, as we currently don't have Playwright tests set up for the offline site. The testing process for the offline site would also be slightly different as it would likely not be tested through a web server, hence why we won't link a Fastly distribution to those buckets. The two online test buckets and Fastly distributions would be in use right away though. If we'd rather wait to create the offline buckets that's fine. If anyone has better suggestions for the bucket names / subdomains feel free to make them as well.
```[tasklist]
### Tasks
- [ ] Update ocw-site Pulumi code to use the new aws.BucketV2 since the old syntax now no longer works
- [ ] Add new S3 buckets as detailed in above design
- [ ] Add Fastly distributions for new buckets
- [ ] Test/validate in CI
- [ ] Roll out QA -> Production
```

### PR DESCRIPTION
**Pull request by feoh**
# What are the relevant tickets?

Fixes #1945

# Description (What does it do?)

Adds end to end ocw testing buckets and Fastly nubbins to match.

Also updates Pulumi code for all S3 bucket creation to the new god awful BucketV2 API.

### CODE REVIEW

**Comment by feoh** [comment]
This apparently broke QA as written.

Please ***DO NOT MERGE***.

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 1]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{draft_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{draft_bucket_name}/images/*",
+                        f"arn:aws:s3:::{draft_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:130]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{draft_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 2]
```diff
@@ -81,161 +83,440 @@
...
                 }
             ],
         }
     ),
-    cors_rules=[{"allowedMethods": ["GET", "HEAD"], "allowedOrigins": ["*"]}],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:104]: This needs to be maintained

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 3]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{draft_backup_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{test_bucket_name}/images/*",
+                        f"arn:aws:s3:::{test_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:184]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{test_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 4]
```diff
@@ -81,161 +83,440 @@
...
                 }
             ],
         }
     ),
-    cors_rules=[{"allowedMethods": ["GET", "HEAD"], "allowedOrigins": ["*"]}],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:125]: We want to maintain the cors rules as well

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 5]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{draft_offline_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{live_bucket_name}/images/*",
+                        f"arn:aws:s3:::{live_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:238]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{live_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 6]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{live_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{draft_backup_bucket_name}/images/*",
+                        f"arn:aws:s3:::{draft_backup_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:292]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{draft_backup_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 7]
```diff
@@ -81,161 +83,440 @@
...
                 }
             ],
         }
     ),
-    cors_rules=[{"allowedMethods": ["GET", "HEAD"], "allowedOrigins": ["*"]}],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:181]: Maintain the cors rules

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 8]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{live_backup_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{live_backup_bucket_name}/images/*",
+                        f"arn:aws:s3:::{live_backup_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:346]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{live_backup_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 9]
```diff
@@ -81,161 +83,440 @@
...
+            "Version": IAM_POLICY_VERSION,
+            "Statement": [
+                {
+                    "Effect": "Allow",
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{draft_offline_bucket_name}/images/*",
+                        f"arn:aws:s3:::{draft_offline_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:400]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{draft_offline_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 10]
```diff
@@ -81,161 +83,440 @@
...
                     "Effect": "Allow",
-                    "Principal": "*",
-                    "Action": ["s3:GetObject"],
-                    "Resource": [f"{live_offline_bucket_arn}/*"],
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{live_offline_bucket_name}/images/*",
+                        f"arn:aws:s3:::{live_offline_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:454]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{live_offline_bucket_arn}/*",
                    ],
```

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 11]
```diff
@@ -81,161 +83,440 @@
...
-    website=s3.BucketWebsiteArgs(
-        index_document="index.html",
-        error_document="error.html",
+    opts=ResourceOptions(
+        depends_on=[
+            live_offline_bucket_public_access,
+            live_offline_bucket_ownership_controls,
+        ]
     ),
-    cors_rules=[{"allowedMethods": ["GET", "HEAD"], "allowedOrigins": ["*"]}],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:236]: Maintain cors rules

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 12]
```diff
@@ -81,161 +83,440 @@
...
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{live_offline_bucket_name}/images/*",
+                        f"arn:aws:s3:::{live_offline_bucket_name}/resumes/*",
+                    ],
                 }
             ],
         }
     ),
-    website=s3.BucketWebsiteArgs(
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:232]: We need to maintain these values as well

File: src/ol_infrastructure/applications/ocw_site/__main__.py [hunk 13]
```diff
@@ -81,161 +83,440 @@
...
+            "Version": IAM_POLICY_VERSION,
+            "Statement": [
+                {
+                    "Effect": "Allow",
+                    "Principal": {"AWS": "*"},
+                    "Action": "s3:GetObject",
+                    "Resource": [
+                        f"arn:aws:s3:::{test_offline_bucket_name}/images/*",
+                        f"arn:aws:s3:::{test_offline_bucket_name}/resumes/*",
+                    ],
```
Comment (blarghmatey) [inline src/ol_infrastructure/applications/ocw_site/__main__.py:508]:
```suggestion
                    "Principal": "*",
                    "Action": "s3:GetObject",
                    "Resource": [
                        f"{test_offline_bucket_arn}/*",
                    ],
```

**Review by blarghmatey — CHANGES_REQUESTED** [review]

File: src/ol_infrastructure/applications/ocw_site/__main__.py (hunk 6 above)
Comment (blarghmatey) [reply src/ol_infrastructure/applications/ocw_site/__main__.py:331]: This bucket still needs the appropriate policy and CORS rules

File: src/ol_infrastructure/applications/ocw_site/__main__.py (hunk 8 above)
Comment (blarghmatey) [reply src/ol_infrastructure/applications/ocw_site/__main__.py:395]: This bucket also still needs the proper policy applied.

**Comment by Ardiea** [comment]
I setup a managed certificate in fastly for https://test-qa.ocw.mit.edu by hand but we should import that into pulumi and figure out how to do the DNS ownership verification automatically. 

https://manage.fastly.com/network/subscriptions/CaYUBzhqebsDvEuZLzATag

And then a CNAME in route53 

_acme-challenge.test-qa.ocw.mit.edu -> zmb1qhi6bx1nvgqsmb.fastly-validations.com

And then the existing CNAME for test-qa.ocw.mit.edu was pointing to j.sni.global.fastly.net but fastly said it should be pointing at d.sni.global.fastly.net so I changed that.

**Review by blarghmatey — APPROVED** [review]

**Comment by gumaerc** [comment]
Sorry for the late comment here. I've been testing the new end to end testing pipeline in QA and everything is working great, but I have one question. In Vault, we have some secrets like `fastly_draft`, `fastly_live` in each environment. Can we add `fastly_test` to that and populate it with an API key like the others? This is important for cache clearing in between runs when there are changes.

**Pull request merged**